utils_nlp/models/transformers/common.py

```diff
@@ -0,0 +1,277 @@
+        else:
+            t_total = len(train_dataloader) // gradient_accumulation_steps * num_train_epochs
+
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": weight_decay,
+            },
+            {
+                "params": [
+                    p for n, p in self.model.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate, eps=adam_epsilon)
+        scheduler = WarmupLinearSchedule(optimizer, warmup_steps=warmup_steps, t_total=t_total)
+
+        if fp16:
+            try:
+                from apex import amp
+            except ImportError:
+                raise ImportError("Please install apex from https://www.github.com/nvidia/apex")
+            self.model, optimizer = amp.initialize(self.model, optimizer, opt_level=fp16_opt_level)
+
+        # multi-gpu training (should be after apex fp16 initialization)
+        if n_gpu > 1:
+            self.model = torch.nn.DataParallel(self.model)
+
+        # Distributed training (should be after apex fp16 initialization)
+        if local_rank != -1:
+            self.model = torch.nn.parallel.DistributedDataParallel(
+                self.model,
+                device_ids=[local_rank],
+                output_device=local_rank,
+                find_unused_parameters=True,
+            )
+
+        global_step = 0
+        tr_loss = 0.0
+        self.model.zero_grad()
+        train_iterator = trange(
+            int(num_train_epochs), desc="Epoch", disable=local_rank not in [-1, 0] or not verbose
+        )
+
+        for _ in train_iterator:
+            epoch_iterator = tqdm(
+                train_dataloader, desc="Iteration", disable=local_rank not in [-1, 0] or not verbose
+            )
+            for step, batch in enumerate(epoch_iterator):
+                self.model.train()
+                batch = tuple(t.to(device) for t in batch)
+                inputs = get_inputs(batch, self.model_name)
+                outputs = self.model(**inputs)
+                loss = outputs[0]
+
+                if n_gpu > 1:
+                    loss = loss.mean()
+                if gradient_accumulation_steps > 1:
+                    loss = loss / gradient_accumulation_steps
+
+                if step % 10 == 0 and verbose:
+                    tqdm.write("Loss:{:.6f}".format(loss / train_batch_size))
+
+                if fp16:
+                    with amp.scale_loss(loss, optimizer) as scaled_loss:
+                        scaled_loss.backward()
+                    torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), max_grad_norm)
```

Comment:
I'm looking at this. I saw different options for this:
- [0.5](https://github.com/yunjey/pytorch-tutorial/blob/master/tutorials/02-intermediate/language_model/main.py#L81)
- [0.25](https://github.com/pytorch/examples/blob/f490f5436958d4fea0250e478b01d0711a09a41a/word_language_model/main.py#L24)
- [1](https://stackoverflow.com/a/54716953)
- [20](https://zhuanlan.zhihu.com/p/76459295)
- [fairseq considers 0](https://github.com/facebookresearch/pytext/blob/master/pytext/optimizer/fp16_optimizer.py#L295) but if we look at the implementation I think there is something weird because having max_norm 0 would make all the [parameters 0](https://pytorch.org/docs/stable/_modules/torch/nn/utils/clip_grad.html)

Reply:
I'm guessing the different values here have minimal impact on the performance, as long as they are within an acceptable range. I don't think we can be sure without some hyperparameter tuning.

Reply:
yeah, makes sense, also we would need to see the profile with tensorboard. We can leave it as 1 and hope that it doesn't mess up the training